lib/IRGen/GenProto.cpp

```diff
@@ -3396,3 +3402,54 @@ llvm::Constant *IRGenModule::getAddrOfGenericEnvironment(
         return fields.finishAndCreateFuture();
       });
 }
+
+//===----------------------------------------------------------------------===//
+// Builtin Protocol Conformances
+//==-----------------------------------------------------------------------===//
+
+static void addTupleEquatableConformance(IRGenModule &IGM,
+                        SmallVectorImpl<ConformanceDescription> &descriptions) {
+  // For now, the only builtin conformance is Equatable for tuples.
+  // (Just grab Void's conformance, we emit a single conformance descriptor for
+  //  all tuples.)
+  auto tuple = IGM.Context.TheEmptyTupleType;
+  auto equatable = IGM.Context.getProtocol(KnownProtocolKind::Equatable);
+  auto conformance = IGM.Context.getBuiltinConformance(tuple, equatable);
+
+  auto description = ConformanceDescription(conformance,
+                                            /* witness table */ nullptr,
+                                            /* pattern */ nullptr,
+                                            /* table size */ 0,
+                                            /* private size */ 0,
+                                            /* requires specialization */ false);
+
+  descriptions.push_back(description);
+}
+
+void IRGenModule::emitBuiltinProtocolConformances() {
+  // We only emit these for the stdlib.
+  if (!getSwiftModule()->isStdlibModule())
+    return;
+
+  // Collect all builtin conformances.
+  SmallVector<ConformanceDescription, 4> descriptions;
+
+  addTupleEquatableConformance(*this, descriptions);
+
+  for (auto description : descriptions) {
+    // Slight edge case: If this conformance doesn't have a valid protocol decl,
+    // we're compiling some module that has asked be to compiled as the stdlib
+    // and said module hasn't declared this protocol. Don't emit descritors for
+    // this case.
+    if (!description.conformance->getProtocol())
+      return;
+
+    addProtocolConformance(std::move(description));
+  }
+}
```

Comment:
Could these data structures be implemented in the runtime C++? It might be easier for people to find and update their implementation that way. (You might have to resort to inline assembly to deal with the relative references though…)

Reply:
They were originally in the runtime until John recommended I move the conformance descriptor into the compiler to remove relative reference issues. Inline assembly is a clever way around that however. I'll look into it because this would clean up a bit of this implementation (this would let me directly reference the witness table now.)

Reply:
Would this require defining the whole structure in inline assembly? Because this might make it a lot less portable and confusing rather than redirecting people to this irgen file to update or add new conformance descriptors.

Reply:
Since we require clang to build the runtime already, the inline assembly format for data should be portable. Another possibility would be to write it as LLVM IR.

Reply:
I believe that I've completed this, but there may be some issues with how I've set it up. I know for a fact that what I have now doesn't work on Windows, but I don't know how to correctly test it. Would love if you took a look and let me know what I need to change about this.